src/typechecker.erl

```diff
@@ -1256,8 +1256,14 @@ expect_tuple_type({type, _, any, []}, _N, _Env) ->
     any;
 expect_tuple_type({type, _, tuple, any}, _N, _Env) ->
     any;
-expect_tuple_type({type, _, tuple, Tys}, N, _Env) when length(Tys) == N ->
-    {elem_ty, Tys, constraints:empty()};
+expect_tuple_type({type, _, tuple, Tys} = Ty, N, _Env) ->
+    Tys = ?assert_type(Tys, [type()]),
+    if
+        length(Tys) == N ->
+            {elem_ty, Tys, constraints:empty()};
+        length(Tys) /= N ->
+            {type_error, Ty}
+    end;
```

Comment:
Let's unpack it:

1. We need to assert that `Tys :: [type()]` to allow for `length(Tys)`. Without that, Gradualizer complains that `Tys :: any | [type()]` and `length(any)` doesn't make sense. A guard failure at runtime just means we proceed to the next clause, but for Gradualizer it's a type error. That's why we have to assert, but assertion is actually a function call, so we have to move `length/1` call from the guard to the body, as arbitrary function calls are not allowed in guards.
2. A line like `Tys = ?assert_type(Tys, [type()])` just updates the var's type in the typechecker env.
3. Since we don't automatically fall through to the next function clause on length mismatch, we have to add the `/= N` `if` clause.

Just so that we're on the same page, I'm not doing these refactorings because "it's fun or they're cool" ;) But because I want to understand how much clunkier Erlang we have to write to satisfy Gradualizer. Therefore, to understand how to improve Gradualizer to enable writing less clunky and more natural Erlang.

The takeaway here is that it would be quite nice if Gradualizer could see that the clause just above this one is:

```erlang
expect_tuple_type({type, _, tuple, any}, _N, _Env) ->
    any;
```

so that `Tys` cannot really ever be `any` 😆 We'll get there one day 🤞

Reply:
Right. Redefining our type()...

Btw, is it really bad style to abuse the exception-in-guard semantics of Erlang? Maybe we're too hard with guards checking and instead can use it to refine the type?

Reply:
> can use it [guards] to refine the type?

Definitely, I think we should do so. This PR makes it even more desirable as we're dealing with unions way more often.

Reply:
In this case though we're guarded against the exception-in-guard, because of the explicit clause matching `{type, _, tuple, any}` anyway.